src/test/parse-fail/impl-trait-plus-priority.rs

```diff
@@ -0,0 +1,47 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// compile-flags: -Z parse-only -Z continue-parse-after-error
+
+fn f() -> impl A + B {} // OK
+fn f() -> dyn A + B {} // OK
+fn f() -> A + B {} // OK
+
+impl S {
+    fn f(self) -> impl A + B { // OK
+        let _ = |a, b| -> impl A + B {}; // OK
+    }
+    fn f(self) -> dyn A + B { // OK
+        let _ = |a, b| -> dyn A + B {}; // OK
+    }
+    fn f(self) -> A + B { // OK
+        let _ = |a, b| -> A + B {}; // OK
+    }
+}
+
+type A = fn() -> impl A + B;
+//~^ ERROR expected a path on the left-hand side of `+`, not `fn() -> impl A`
+type A = fn() -> dyn A + B;
+//~^ ERROR expected a path on the left-hand side of `+`, not `fn() -> dyn A`
+type A = fn() -> A + B;
+//~^ ERROR expected a path on the left-hand side of `+`, not `fn() -> A`
+
+type A = Fn() -> impl A + B; // OK, interpreted as `(Fn() -> impl A) + B`
+type A = Fn() -> dyn A + B; // OK, interpreted as `(Fn() -> dyn A) + B`
```

Comment:
But this case probably doesn't error -- I sort of think it should.

Reply:
I suppose we might be able to target it by linting. I did think though there was some desire to "put off" the question of just how `Fn` and friends works when combined with `dyn` and `impl`.

Reply:
+1-- I think users should have to manually specify parentheses here, at least for the time-being.

Reply:
It'll be a bit of a pain to do, I guess, since saying `dyn A + B` and `A + B` have same priority means that this case just kind of "falls out" this way. 

But I still feel like we should keep future flexibility here. 

@petrochenkov what do you think?

Reply:
Clippy lint?

To put this into perspective - few people remember priorities of operators in expressions beyond `+` vs `*` and `&&` vs `||`, but still don't require parentheses everywhere.
Here if you get the priorities of `Fn` and `+` wrong, you won't even get runtime bugs, like with expressions, you'll get a type checking error about unsatisfied bounds, giving you opportunity to learn what priority is right.

Reply:
We'd need to be conservative if we were going to change the priority, but that's not the case.
For prefix-less `+` and `dyn` the current priority is always correct because you can't return unsized values from functions, for `impl` it's fifty-fifty and misinterpretations are possible with both rules, so the priority is generally correct and there's no need to change it.

Reply:
The lang team discussed this at a previous meeting and were unable to reach a conclusion. Most felt that `Fn() -> dyn A + B` should either parse as `Fn() -> (A + B)` or not at all. Can you modify the PR to error in this case so that we can have more flexibility going forwards?

Reply:
More directly, I think the point was that we **know** we want this to work:

```rust
fn foo() -> impl A + B { }
```

and that therefore one might very well expect this to work:

```rust
where F: Fn() -> impl A + B
```

Now, historically, as @petrochenkov correctly points out, we opted *not* to make the where clause case "work" (that is, `where F: Fn() -> A + B` parses two bounds on `F`) because returning a value of type `A + B` was almost certainly not what you wanted. However, introducing keywords like `dyn` and `impl` gives us a chance to revisit this decision backwards compatibly, and it is not clear that -- in this new context -- this is the correct behavior.

As @cramertj noted, I don't think that *anyone* feels the correct interpretation of `where F: Fn() -> impl A + B` would be that `B` is a bound on `F`. (It's just a bit too surprising.) And moreover it seems clear that `dyn`  ought to behave the same as `impl`. However, there were also those -- notably @joshtriplett -- who felt that there is *no* correct interpretation, and we should just avoid guessing. (I think that @joshtriplett would also be happy to have required parentheses in other cases where we do not today, e.g. amongst confusing operators.)

Given this, it does seem like we ought to try to disallow `-> impl A + B` and `-> dyn A + B` for now.

Reply:
Let me try to clarify what I think we want. We already have two grammatical precedence levels for types. I'll them `T0` and `T1`. `T0` includes all types, including the "n-ary" operators like `dyn A + B` and `A + B`. `T1` excludes those n-ary operators, and is used primarily for `&` types (which take a `T1` argument). 

Currently, in the parser, if we are parsing a `T1` type and we see a `+` we stop, leaving it to be consumed from some enclosing context.

I think that we want to have a rule such that `T1 = dyn Identifier` parses, but if we see a `+` after that, we error out (in the parser).  This means that `&dyn A + B` is a parse error, as is `where F: Fn() -> dyn A + B`, and (I imagine) `dyn Fn() -> dyn A + B`.

I think of this as being analogous to non-associativity: there are basically multiple interpretations of the `+` operator. It can be adding bounds to a type parameter in a where-clause list; it can be adding types in a (old-style, `A+B`) sum-type; and it can be adding bounds to a `dyn` or `impl` type.

In a fn item context, or in the `T` in `Box<T>`, once we see `dyn` or `impl`, there is only one valid interpretation (since we are not in a where clause list) -- . These are exactly the cases (I believe) where we use the grammatical nonterminal `T0`. We can parse eagerly there.

In other contexts, there are multiple contending interpretations. We are aiming not to choose between them (hence "non-associative").

Reply:
@nikomatsakis 
Thanks for the detailed write up, this is better than just "lang team decided so".
I agree that consistency with `fn f() -> A + B { ... }` is a good argument in favor of swapping priorities for `Fn` and `+`.
I'll implement the conservative "refuse to disambiguate" solution.

Reply:
This discussion should've probably been happening in https://github.com/rust-lang/rfcs/pull/2250, because the RFC needs to be updated as well.

Reply:
Agreed, I'll copy my post over there.